Wire up debug events

briandealwis opened the pull request:
Relates to #3122 (#3564, #3609)
Fixes #2211 

**Description**

This PR wires up the new `DebuggingContainerEvent` (#3609) to be fired by Skaffold when observing a debuggable container starting or terminating. These debugging containers are exposed via the Skaffold state object. With this PR, the IDEs can monitor for events to populate UIs.

**User facing changes**

New events and addition to the Skaffold state.

There are no changes to the log.  Although some logging may be useful for users seeking to configure their debuggers manually, I figured the port-forwarding messages would usually be sufficient.

<details>
<summary>`/v1/events` stream of `skaffold debug` within `examples/jib`</summary>

In this example, we do a `skaffold debug`, and then kill the deployed pod.  The deployment starts a new pod.  We get a terminated event for the container for the killed pod.

```json
{
  "result": {
    "timestamp": "2020-02-05T03:27:30.114354Z",
    "event": {
      "debuggingContainerEvent": {
        "status": "Started",
        "podName": "web-f6d56bcc5-6csgs",
        "containerName": "web",
        "namespace": "default",
        "artifact": "skaffold-jib",
        "runtime": "jvm",
        "debugPorts": {
          "jdwp": 5005
        }
      }
    },
    "entry": "Debuggable container started pod/web-f6d56bcc5-6csgs:web (default)"
  }
}
```

</details>

<details>
<summary>The `/v1/state` listing debugging containers</summary>

```json
{
  "buildState": {
    "artifacts": {
      "skaffold-jib": "Complete"
    }
  },
  "deployState": {
    "status": "Complete"
  },
  "forwardedPorts": {
    "5005": {
      "localPort": 5005,
      "remotePort": 5005,
      "podName": "web-f6d56bcc5-6csgs",
      "containerName": "web",
      "namespace": "default",
      "portName": "jdwp",
      "resourceType": "pod",
      "resourceName": "web-f6d56bcc5-6csgs",
      "address": "127.0.0.1"
    },
    "8080": {
      "localPort": 8080,
      "remotePort": 8080,
      "namespace": "default",
      "resourceType": "service",
      "resourceName": "web",
      "address": "127.0.0.1"
    },
    "8081": {
      "localPort": 8081,
      "remotePort": 8080,
      "podName": "web-f6d56bcc5-6csgs",
      "containerName": "web",
      "namespace": "default",
      "resourceType": "pod",
      "resourceName": "web-f6d56bcc5-6csgs",
      "address": "127.0.0.1"
    }
  },
  "statusCheckState": {
    "status": "Not Started"
  },
  "fileSyncState": {
    "status": "Not Started"
  },
  "debuggingContainers": [
    {
      "status": "Started",
      "podName": "web-f6d56bcc5-6csgs",
      "containerName": "web",
      "namespace": "default",
      "artifact": "skaffold-jib",
      "runtime": "jvm",
      "debugPorts": {
        "jdwp": 5005
      }
    }
  ]
}

```

</details>



**Next PRs.**

<!-- In this section describe a list of follow up PRs if the current PR is a part of big feature change.
See example https://github.com/GoogleContainerTools/skaffold/pull/2811
Write n/a if not applicable.
-->


**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [x] Mentions any output changes.
- [ ] ~Adds documentation as needed: user docs, YAML reference, CLI reference.~ _already done_
- [x] Adds integration tests if needed.

<!--
_See [the contribution guide](../CONTRIBUTING.md) for more details._
Double check this list of stuff that's easy to miss:
- If you are adding [a example to the `examples` dir](https://github.com/GoogleContainerTools/skaffold/tree/master/examples), please copy them to [`integration/examples`](https://github.com/GoogleContainerTools/skaffold/tree/master/integration/examples)
- Every new example added in [`integration/examples` dir](https://github.com/GoogleContainerTools/skaffold/tree/master/integration/examples), should be tested in [integration test](https://github.com/GoogleContainerTools/skaffold/tree/master/integration)
-->

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit test added.
- [ ] User facing changes look good.


**Release Notes**

<!-- 
Describe any user facing changes here so maintainer can include it in the release notes, or delete this block.
-->

```
- New events emitted when debuggable containers are started and terminated
```

briandealwis commented:
One point to highlight: I introduced a `SkaffoldRunner.isDebugMode()` method that tests whether Skaffold was launched as `skaffold debug`.  This works for debug, but it doesn't seem a great pattern — for example, `debug` effectively invokes `dev` under the hood.

briandealwis commented on pkg/skaffold/runner/debugging.go:
```diff
@@ -0,0 +1,41 @@
+// isDebugMode returns true if we're running for debugging.
+func (r *SkaffoldRunner) isDebugMode() bool {
+	return r.runCtx.Opts.Command == "debug"
+}
```
I introduced this function to test whether Skaffold was launched as `skaffold debug`. This works for debug, but it doesn't seem a great pattern — for example, debug effectively invokes dev under the hood.

balopat replied:
nit: we already have a `runContext.DevMode` that serves a similar purpose. Can we make this similar? I don't have a preference towards either, I just don't want inconsistency.

briandealwis commented on pkg/skaffold/kubernetes/debugging/container_manager.go:
```diff
@@ -0,0 +1,151 @@
+	namespaces  []string
+	active      map[string]string // set of containers that have been notified
+	aggregate   chan watch.Event
+}
```
I had originally hoped to merge this into the port-forwarding `ForwarderManager`, but since there are some language runtimes that don't actually use forwarded ports (they instead require `kubectl ssh` and operate across stdin/stdout), I introduced this new `ContainerManager`.  The lifecycle of a debugging container doesn't match forwarded ports either.

balopat replied:
Which language runtimes require stdin/stdout?

briandealwis replied:
.NET

briandealwis replied:
.NET

briandealwis commented on pkg/skaffold/kubernetes/debugging/container_manager.go:
```diff
@@ -0,0 +1,151 @@
+				// "DELETED" if the pod is deleted
+				if evt.Type != watch.Modified && evt.Type != watch.Deleted {
+					continue
+				}
```
What I observed when prototyping this functionality that I couldn't rely on the pod status.  Rather I had to check on pod MODIFIED and DELETED events and then look at and track the individual container status (see below in `checkPod()`).  I would get spurious debug events otherwise.

balopat commented on pkg/skaffold/kubernetes/debugging/container_manager.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+func (d *ContainerManager) Start(ctx context.Context) error {
+	if d == nil {
```
yeah this would be the 3rd podWatcher based control loop in skaffold with the exact same control flow. 

I would think a single podWatcher would be less resource intensive (the same "event messages" only need to be sent once for skaffold) on the API server side and on the skaffold side as well. 

I think, in a separate PR we should refactor first to a "visitor" pattern on a centralized podWatcher and make the logs, portforwards and debug listeners "register" with the podWatcher. 

In this visitor framework the implementation of the debug listener only would require line 94-99.

balopat replied:
OK. As discussed, we can push this now and refactor in a follow-up PR.

balopat commented:
Hmm...the new test just flaked: 

```
time="2020-02-27T20:10:21Z" level=info msg="Ran in 9.814053538s"
--- FAIL: TestDebugEventsRPC (13.69s)
    rpc_test.go:376: waiting for connection...
    rpc_test.go:376: waiting for connection...
    debug_test.go:126: int differ (-got, +want):   int(
        - 	0,
        + 	1,
          )
```

balopat commented on integration/debug_test.go:
```diff
@@ -71,3 +74,54 @@ func TestDebug(t *testing.T) {
+
+	metaEntries, buildEntries, deployEntries, debuggingEntries := 0, 0, 0, 0
+	timeout := time.After(1 * time.Minute)
+	for i := 0; i < 6; i++ {
```
this is brittle, what if we change events? 
it also did fail with the new status events introduced

balopat replied:
I did change it.

briandealwis commented:
So the cause is that I deviate from [port-forwarding's watch event filtering](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/kubernetes/portforward/pod_forwarder.go#L87-L90), which looks at ADDED and MODIFIED events, whereas the container manager [only looks at MODIFIED and DELETED watch events](https://github.com/GoogleContainerTools/skaffold/pull/3645/files#diff-0fcbb903815f224038a7c6a0c80e81aeR90-R96).  Prior to the status checks, ADDED events seemed spurious as they had no `ContainerStatus` items — the real changes came through MODIFIED events.

Removing this watch event filtering and using only the container status makes everything work again.

briandealwis commented:
PTAL @balopat
